kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```diff
@@ -227,15 +227,42 @@ sealed abstract class Resource[+F[_], +A] {
    */
   def mapK[G[x] >: F[x], H[_]](
       f: G ~> H
-  )(implicit D: Defer[H], G: Applicative[H]): Resource[H, A] =
-    this match {
-      case Allocate(resource) =>
-        Allocate(f(resource).map { case (a, r) => (a, r.andThen(u => f(u))) })
-      case Bind(source, f0) =>
-        Bind(Suspend(D.defer(G.pure(source.mapK(f)))), f0.andThen(_.mapK(f)))
-      case Suspend(resource) =>
-        Suspend(f(resource).map(_.mapK(f)))
-    }
+  )(implicit G: Applicative[H]): Resource[H, A] = {
```

Comment:
It looks to me that you can weaken this to `Functor`.
There is also a choice to be made here - should the constraint be on `G` or `H`?
```suggestion
  )(implicit H: Functor[H]): Resource[H, A] = {
```

Reply:
> It looks to me that you can weaken this to `Functor`.

`Applicative` is needed for the `map` calls.

> There is also a choice to be made here - should the constraint be on `G` or `H`?

It needs to be on `H` for the `flatMap` calls - I don't think those can be done in `G`.

Reply:
Oh - `flatMap` doesn't require any instance, so you're right, we could move the constraint from `H` to `G`. Do you see any particular benefit to having it on `G` rather than `H`?

Reply:
> Applicative is needed for the map calls.

`Functor` gives you `map`

> Do you see any particular benefit to having it on G rather than H?

I don't have any arguments in either direction. Maybe that's just an arbitrary choice 🤔

Reply:
> `Functor` gives you `map`

You're right - my bad, I was thinking the `map` calls were on `Resource`.